service/matching/matching_engine.go

```diff
@@ -371,60 +371,129 @@ func (e *matchingEngineImpl) getTaskQueuePartitionManager(
 	create bool,
 	loadCause loadCause,
 ) (taskQueuePartitionManager, error) {
-	pm, err := e.getTaskQueuePartitionManagerNoWait(partition, create, loadCause)
+	key := partition.Key()
+
+	e.partitionsLock.RLock()
+	pm, ok := e.partitions[key]
+	e.partitionsLock.RUnlock()
+
+	if ok {
+		return pm, nil
+	}
+
+	if !create {
+		return nil, nil
+	}
+
+	pm, created, err := e.createTaskQueuePartitionManager(partition, loadCause)
 	if err != nil || pm == nil {
 		return nil, err
 	}
+
+	if pm.Partition().IsRoot() && created {
+		// Whenever a root partition is loaded we need to force all other partitions to load.
+		// If there is a backlog of tasks on any child partitions force loading will ensure that they
+		// can forward their tasks the poller which caused the root partition to be loaded.
+		// These partitions could be managed by this matchingEngineImpl, but are most likely not.
+		// We skip checking and just make gRPC requests to force loading them all.
+		err := e.forceLoadAllNonRootPartitions(ctx, partition)
+		if err != nil {
+			e.logger.Error("Failed to force load all non-root partitions after root partition was loaded",
+				tag.WorkflowNamespaceID(partition.NamespaceId().String()),
+				tag.WorkflowTaskQueueName(partition.TaskQueue().Name()),
+				tag.WorkflowTaskQueueType(partition.TaskType()),
+				tag.Error(err))
+		}
+	}
+
 	if err = pm.WaitUntilInitialized(ctx); err != nil {
 		e.unloadTaskQueuePartition(pm, unloadCauseInitError)
 		return nil, err
 	}
 	return pm, nil
 }
 
-// Returns taskQueuePartitionManager for a task queue. If not already cached, and create is true, tries
-// to get new range from DB and create one. This does not block for the task queue to be initialized.
-func (e *matchingEngineImpl) getTaskQueuePartitionManagerNoWait(
+func (e *matchingEngineImpl) forceLoadAllNonRootPartitions(
+	ctx context.Context,
+	rootPar tqid.Partition,
```

Comment:
```suggestion
	rootPartition tqid.Partition,
```
or just `root` if you want. avoid abbreviations

Reply:
Done.